### PR TITLE
feat: lm-evaluation-harness network compatibility for vMLX server

### PR DESCRIPTION
**Pull request by SandorDobi**
## Summary

This PR makes vMLX's OpenAI-compatible server a **drop-in backend** for [EleutherAI/lm-evaluation-harness](https://github.com/EleutherAI/lm-evaluation-harness) (v0.4.11+) via the `local-completions` adapter, using **only HTTP requests** — no local HuggingFace weights or tokenizer required.

**Validation: 49/49 behavioral assertions passed** across 7 milestones, including end-to-end `lm_eval` runs for `hellaswag`, `gsm8k`, and `record` tasks.

---

### 1. Tokenizer Endpoints (3 new routes)

| Endpoint | Method | Description |
|----------|--------|-------------|
| `/v1/tokenizer_info` | GET | `eos_token`, `bos_token`, `pad_token`, `chat_template` |
| `/v1/tokenize` | POST | `"text"` → `[token_ids]` |
| `/v1/detokenize` | POST | `[token_ids]` → `"text"` |

Also registers `/tokenizer_info`, `/tokenize`, `/detokenize` aliases (lm-eval strips `/v1/completions` from `base_url`).

### 2. Legacy Completions Logprobs

Full OpenAI **legacy completions logprobs format** for `/v1/completions`:
- `tokens`, `token_logprobs`, `top_logprobs`, `text_offset`
- Works in both **streaming (SSE)** and **non-streaming** modes
- Handles `logprobs: true` and `logprobs: 1` (integer, as lm-eval sends)

### 3. Echo + Prompt Logprobs

- `echo: true` — returns prompt text + logprobs for every prompt token
- `prompt_logprobs: N` — returns prompt logprobs without echoing text
- **Batched `logSoftmax`** over full `[1, seq_len, vocab]` tensor (not per-token)
- Position 0 → `null` (no prior context)
- Gated behind explicit params: **zero overhead for normal requests**

### 4. Loglikelihood Fast-Path

Dedicated synchronous path for `max_tokens: 1` loglikelihood requests that bypasses:
- `AsyncStream` / `continuation.yield`
- `StreamingDetokenizer`
- `ToolCallProcessor`

Measured at **103% of raw prefill latency** (3% overhead, target was <150%).

### 5. RawPrompt Mode

Legacy `/v1/completions` requests skip the chat template, ensuring token counts match lm-eval's `ctxlen` for correct `parse_logprobs()` slicing.

### 6. Prompt-as-[Int] Support

`/v1/completions` accepts pre-tokenized prompts:
- `prompt: "string"` — existing
- `prompt: [128000, 128006]` — token ID array (lm-eval `tokenized_requests=True`)
- `prompt: [[128000], [128006]]` — batched (joins with `\n`)

### 7. lm-eval Integration Verified

```bash
# generate_until tasks
lm-eval --model local-completions --model_args \
  "base_url=http://localhost:8080/v1/completions" \
  --tasks gsm8k --limit 2 --batch_size 1

# loglikelihood tasks with remote tokenizer
lm-eval --model local-completions --model_args \
  "base_url=http://localhost:8080/v1/completions,tokenizer_backend=remote" \
  --tasks hellaswag --limit 5 --batch_size 1
```

All exit code 0 with valid metrics.

---

## Architecture Highlights

```
Normal chat completion (no logprobs):
  Client → OpenAIRoutes → ChatRequest → Engine.stream → TokenIterator
    → model.prepare() → convertToToken() → generateLoopTask
    → AsyncStream → SSEEncoder → Client

Loglikelihood fast-path:
  Client → OpenAIRoutes → ChatRequest(maxTokens:1, echo, logprobs)
    → Engine.loglikelihood() → model() → logSoftmax → response
    → NO AsyncStream, NO StreamingDetokenizer, NO ToolCallProcessor
```

**Key invariants:**
1. Normal requests: **zero overhead** — no `logSoftmax` when not requested
2. Prompt logprobs: **batched** single `logSoftmax` over full sequence
3. Top-K: **`argPartition`** (O(V)) not `argSort` (O(V log V))
4. Float32: `bfloat16` logits upcast before `logSoftmax` (no NaN/Inf)
5. KV cache bypass: prompt logprobs skip cache to ensure fresh prefill

---

## Files Changed

| File | What changed |
|------|-------------|
| `Sources/vMLXLMCommon/Evaluate.swift` | `LogprobsCollector`, prompt logprob capture, `argPartition`, `Generation.logprob` |
| `Sources/vMLXEngine/Engine.swift` | `Engine.loglikelihood()`, `LoglikelihoodResult`, fast-path |
| `Sources/vMLXEngine/ChatRequest.swift` | `echo`, `promptLogprobs`, `rawPrompt` |
| `Sources/vMLXEngine/Stream.swift` | `flushLogprobs()` before tool calls |
| `Sources/vMLXServer/Routes/OpenAIRoutes.swift` | Tokenizer routes, completions logprobs, legacy format, prompt-as-[Int] |
| `Sources/vMLXServer/SSEEncoder.swift` | `buildLegacyLogprobs()`, `buildLegacyLogprobsJSON()`, ordered dict serialization |
| `Package.swift` | `vMLXServer` → `vMLXLMCommon` dependency |
| `tests/e2e/harness.sh` | New suites: tokenizer, completions, logprobs, fastpath, legacy |
| `docs/LM-EVAL-HARNESS-*.md` | Compatibility plan, implementation summary |

---

## Validation

| Milestone | Assertions | Status |
|-----------|-----------|--------|
| Tokenizer Endpoints | 7/7 | ✅ |
| Completions Logprobs | 7/7 | ✅ |
| Echo + Prompt Logprobs | 11/11 | ✅ |
| Legacy Logprobs Format | 10/10 | ✅ |
| Loglikelihood Fast-Path | 3/3 | ✅ |
| Numerical Correctness + Chat | 3/3 | ✅ |
| lm-eval Integration | 8/8 | ✅ |
| **Total** | **49/49** | **✅** |

---

## Known Limitations

1. **Batching:** `/v1/completions` joins `[String]` prompts with `\n`. Use `batch_size=1` in lm-eval.
2. **pad_token:** Uses `unknownToken` as stand-in (Tokenizer protocol lacks `padToken`).
3. **chat_template:** Returns `null` when raw template string is unavailable.
4. **Streaming chat template:** Streaming `max_tokens>1` still applies chat template (rawPrompt only affects fast-path).
5. **Multi-step prefill:** Models returning `.tokens` from `model.prepare()` don't support prompt logprobs (most models return `.logits`).

---

## Related

- Original logprobs implementation: iter-96 §123
- Compatibility plan: `docs/LM-EVAL-HARNESS-COMPATIBILITY.md`
- Implementation summary: `docs/LM-EVAL-HARNESS-IMPLEMENTATION-SUMMARY.md`

### CODE REVIEW

**Comment by jjang-ai** [comment]
Thanks for this work — it's a substantial contribution and the architecture (batched `logSoftmax` + `argPartition` top-K + `Engine.loglikelihood()` synchronous fast-path bypassing `AsyncStream`/`StreamingDetokenizer`/`ToolCallProcessor`) is the right shape for lm-evaluation-harness.

The PR has been sitting blocked on `CONFLICTING` for a while because `dev` moved ~30 commits forward (§221 → §346) since the fixup branch was cut. To unblock it I've:

1. Cherry-picked `b018159` (your iter-96 §123) → `a638cf7` (handoff docs) → `8ad1850` (the §163 B1–B6 fixups: `bytes: [Int]` field, argPartition top-K, stop-sequence doc, merged-frame logprobs, legacy /v1/completions, rollback scenarios) onto current `origin/dev` (HEAD `ca4bf63`).
2. Resolved the one textual conflict in `ChatRequest.swift` — dropped the pre-iter-96 "reject logprobs" guard since your branch implements it, preserved the §173 `frequency_penalty`/`presence_penalty` wiring note.
3. Auto-merged Package.swift, Stream.swift, Evaluate.swift, OpenAIRoutes.swift, SSEEncoder.swift, tests/e2e/harness.sh — no semantic conflicts with recent dev work (§297 x-vmlx-trace-id header, §330 stream_options passthrough, §262 backpressure, §344 parser-drain EOS fix, §346 JANGTQ mxtq_bits decode all orthogonal to the logprobs surface).
4. Verified `swift build` green (1999 artifacts, 48.9s) and `swift test` on the relevant filter (ChatRequestValidationTests, APISurfaceRegressionTests, SettingsStoreTests, JANGTQConfigBitsDecodingTests, ChatTemplateReproTests, RealReasoningDispatchTests): 100 pass / 1 stale fail — `testRejectsLogprobsTrue` still asserts the old reject behavior; that test needs inverting now that logprobs are accepted.

The rebased branch is pushed as [`pr99-rebased-onto-dev`](https://github.com/jjang-ai/vmlx/tree/pr99-rebased-onto-dev). I'm going to merge it into `dev` from there and close this PR — your commits (b018159, a638cf7, 8ad1850) are preserved verbatim in that integration branch, so the `git log` will continue to show your authorship on the feature.

Follow-ups I'm going to pick up in-house:
- Invert the stale test
- Close the 5 "Known Limitations" you noted (especially the `pad_token` / `chat_template: null` gaps on `/v1/tokenizer_info` and the multi-step prefill prompt_logprobs gap)
- Live-verify the fast-path against a real lm-eval run

Closing when the merge lands. Thanks again — this is a meaningful step toward lm-eval compatibility.

**Comment by jjang-ai** [comment]
Merged into dev via the rebased integration branch `pr99-rebased-onto-dev` (commits 1ce0f5a, 2579832, cb01f95 — your authorship preserved verbatim). See PR comment above for rebase details. Thank you for the contribution — picking up the in-house follow-ups now.

**Comment by jjang-ai** [comment]
Crediting you in **v1.3.85** release notes — your lm-eval-harness network compatibility work (the OpenAI-compatible logprobs / top_logprobs plumbing in this PR) is the foundation that §347 (remote tokenizer HTTP routes) and §352 (legacy /v1/completions logprobs wire) built on top of. Thanks @SandorDobi 🙏

Release: https://github.com/jjang-ai/mlxstudio/releases/tag/v1.3.85
